### PR TITLE
Expose Fretboard bucket # and add bucket distribution test

### PR DESCRIPTION
**Pull request by colintheshots**

### CODE REVIEW

**Comment by Amejia481** [comment]
Hi @colintheshots! 
Could you help me to clarify what is the goal of this PR?

File: components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt [hunk 1]
```diff
@@ -431,4 +433,59 @@ class FretboardTest {
...
+            f.call(experimentEvaluator, context) as Int
+        }
+
+        println("Into deciles")
+        distribution
+                .groupingBy { it / 10 }
+                .eachCount()
+                .toSortedMap()
+                .forEach {
+                    println("${it.key * 10}..${it.key * 10 + 9} : ${it.value}")
```
Comment (Amejia481) [inline components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt]: Do you want to keep the print lines?

File: components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt [hunk 2]
```diff
@@ -431,4 +433,59 @@ class FretboardTest {
...
+                    assertTrue(it.value in 75..125)
+                }
+
+        println("Into halves")
+        distribution
+                .groupingBy { it / 50 }
+                .eachCount()
+                .toSortedMap()
+                .forEach {
+                    println("${it.key * 500}..${it.key * 500 + 499} : ${it.value}")
```
Comment (Amejia481) [inline components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt]: Do you want to keep the print lines?

**Comment by colintheshots** [comment]
This PR exists because if our data group cannot figure out which bucket users are meant to be in, it's really hard to diagnose telemetry issues. See focus-android#3443

Also, we were asked explicitly if any tests existed to confirm the proper bucket distribution in Fretboard, so I wrote one.

I can remove the print statements.

File: components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt [hunk 3]
```diff
@@ -431,4 +433,57 @@ class FretboardTest {
...
+        println("Into deciles")
+        distribution
+                .groupingBy { it / 10 }
+                .eachCount()
+                .toSortedMap()
+                .forEach {
+                    assertTrue(it.value in 75..125)
+                }
+
+        println("Into halves")
```
Comment (Amejia481) [inline components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt]: This one too!

File: components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt [hunk 4]
```diff
@@ -431,4 +433,57 @@ class FretboardTest {
...
+        `when`(context.getSharedPreferences(ArgumentMatchers.anyString(), ArgumentMatchers.anyInt())).thenReturn(sharedPrefs)
+
+        val distribution = (1..1000).map {
+            val experimentEvaluator = ExperimentEvaluator()
+            val f = experimentEvaluator::class.functions.find { it.name == "getUserBucket" }
+            f!!.isAccessible = true
+            f.call(experimentEvaluator, context) as Int
+        }
+
+        println("Into deciles")
```
Comment (Amejia481) [inline components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt]: Sorry, I missed this one!

File: components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt [hunk 5]
```diff
@@ -431,4 +433,55 @@ class FretboardTest {
         val fretboard = Fretboard(source, storage)
         fretboard.updateExperiments()
     }
+
+    @Test
+    fun getUserBucketTest() {
```
Comment (Amejia481) [inline components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt]: Sorry, just a small one!
Can we make this test name standard as the other ones.
`testUserBucket(`)

File: components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt [hunk 6]
```diff
@@ -431,4 +433,55 @@ class FretboardTest {
...
+        `when`(prefsEditor.putString(ArgumentMatchers.anyString(), ArgumentMatchers.anyString())).thenReturn(prefsEditor)
+        `when`(context.getSharedPreferences(ArgumentMatchers.anyString(), ArgumentMatchers.anyInt())).thenReturn(sharedPrefs)
+        `when`(sharedPrefs.getString(ArgumentMatchers.anyString(), ArgumentMatchers.isNull()))
+                .thenReturn("a94b1dab-030e-4b13-be15-cc80c1eda8b3")
+        val fretboard = Fretboard(experimentSource, experimentStorage)
+        assertTrue(fretboard.getUserBucket(context) == 54)
+    }
+
+    @Test
+    fun evenDistributionTest() {
```
Comment (Amejia481) [inline components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt]: Also this one.
`testEvenDistribution()`

**Comment by Amejia481** [comment]
Thanks @colintheshots!

**Comment by csadilek** [comment]
I will take a look at this as well.

**Comment by colintheshots** [comment]
Not being a mathematician, I don't know what reasonable discrepancy from an ideal distribution is acceptable. I was aiming to validate whether or not uneven distribution was related to our problems in beta. I don't believe it's the issue, but we should have a valid test.

**Comment by csadilek** [comment]
@colintheshots understood. Glad you wrote the test! Just need to make sure the upper and lower bounds of the assertion aren't causing intermittent failures. So, we need to understand what to set them to :). Right now [70, 130] is causing failures for me. Keep hitting 131.

**Comment by csadilek** [comment]
OK, we have https://github.com/mozilla-mobile/android-components/issues/809, which I will keep open for improving the tests. We'll want to implement this similar to: https://github.com/mozilla-mobile/firefox-ios/blob/master/SharedTests/FeatureSwitchTests.swift#L139 in a coming sprint.

**Pull request merged**